ranger/core/actions.py

```diff
@@ -294,7 +307,12 @@ def substitute_macros(self, string,  # pylint: disable=redefined-outer-name
             raise ValueError("Could not apply macros to `%s'" % string)
         return result
 
-    def get_macros(self):  # pylint: disable=too-many-branches,too-many-statements
+    def get_macros(self):
```

Comment:
What does the indirection to `_get_macros` buy us? Are you just memoizing the call, maybe a decorator is more appropriate?

Reply:
Yes, but the function `get_macros`existed before that and was returning the dict of all macros. The function can be used in plugins so I assume it is used by people, i.e. it is used by code that we do not control. The macros dict is, however, meant to be generated only once - because of the whole lazyness that we want. Originally I had the memoization in the `substitute_macros` function but this leaves the problem that `get_macros` can be called from the outside world, thus generating the dict again. Putting the dict generation in `_get_macros` lets us keep the "generate dict only once" idea  even if the function `get_macro` is called from wherever.

Reply:
I still don't understand this tbh. You're *just* memoizing `get_macros`, right? And the only place you call `_get_macros` is in `get_macros`? So I still think a memoizing decorator is the way to go, it's more explicit about what's going on *and* more reusable.

Reply:
Does this memoizing decorator exists or do we have to write it?

Reply:
We don't have one yet, so you may have to define one.